src/github.com/matrix-org/dendrite/clientapi/writers/register.go

```diff
@@ -0,0 +1,153 @@
+package writers
+
+import (
+	"net/http"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/dendrite/clientapi/auth/storage"
+	"github.com/matrix-org/dendrite/clientapi/auth/types"
+	"github.com/matrix-org/dendrite/clientapi/httputil"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+)
+
+// registerRequest represents the submitted registration request.
+// It can be broken down into 2 sections: the auth dictionary and registration parameters.
+// Registration parameters vary depending on the request, and will need to remembered across
+// sessions. If no parameters are supplied, the server should use the parameters previously
+// remembered. If ANY parameters are supplied, the server should REPLACE all knowledge of
+// previous paramters with the ones supplied. This mean you cannot "build up" request params.
+type registerRequest struct {
+	// registration parameters.
+	Password string `json:"password"`
+	Username string `json:"username"`
+	// user-interactive auth params
+	Auth authDict `json:"auth"`
+}
+
+type authDict struct {
+	Type    types.LoginType `json:"type"`
+	Session string          `json:"session"`
+	// TODO: Lots of custom keys depending on the type
+}
+
+// http://matrix.org/speculator/spec/HEAD/client_server/unstable.html#user-interactive-authentication-api
+type userInteractiveResponse struct {
+	Flows     []flow                 `json:"flows"`
+	Completed []types.LoginType      `json:"completed"`
+	Params    map[string]interface{} `json:"params"`
+	Session   string                 `json:"session"`
+}
+
+type flow struct {
```

Comment:
Called it `authFlow` since it's not just for login, but also registration. https://github.com/matrix-org/dendrite/pull/113/commits/9cae3887db6e477d6b446dce41a680c545527026